kdf/kdf.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2019 Thales e-Security, Inc
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+package kdf
+
+import (
+	"crypto"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"io"
+
+	// Force registration of SHA1 and SHA2 families of cryptographic primitives to
+	// reduce the burden on KDF consuming packages.
+	_ "crypto/sha1"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+)
+
+var (
+	// errInvalidLengthParameter the KDF length parameter is invalid
+	errInvalidLengthParameter = errors.New("invalid length parameter")
+
+	// errInvalidSeedParameter a parameter is invalid.
+	errInvalidSeedParameter = errors.New("invalid input parameter")
+)
+
+// Verify KDF completely implements the io.Reader interface.
+var _ io.Reader = (*KDF)(nil)
+
+// KDF key derivation context struct
+type KDF struct {
+	seed       []byte
+	other      []byte
+	length     int
+	iterations uint32
+	position   int
+	buffer     []byte
+	digester   hash.Hash
+}
+
+// i2osp 4-byte integer marshalling.
+func i2osp(i uint32) []byte {
+	osp := make([]byte, 4)
+	binary.BigEndian.PutUint32(osp, i)
+	return osp
+}
+
+// min select the minimum value of a and b.
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// Read read the next len(p) bytes from the KDF context.
+func (kdf *KDF) Read(p []byte) (int, error) {
+	var n int
+	// When there's no data left return EOF.
+	if kdf.length-kdf.position == 0 {
+		return 0, io.EOF
```

Comment:
If someone passes an empty slice for `p`, I think we should return `0, nil`, regardless of whether the KDF is empty. The `io.Reader` docs don't go as far as mandating this, but they say:

>Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0.

Reply:
This is now address in 9cc1e4096b7ce918fdb68b229c195fe819657418 whereby if `len(p)` is 0,  the function will return as if it would if it was satisfied from copying buffered data.